File: src/main/java/no/nav/helse/flex/operations/SkjemaMetadata.java
package no.nav.helse.flex.operations;

import com.google.common.reflect.TypeToken;
import com.google.gson.Gson;
import no.nav.helse.flex.Environment;

import java.util.HashMap;
import java.util.List;
import java.util.Objects;

public class SkjemaMetadata {
    private final HashMap<String, TemaKodeverk> temaMap;

    public SkjemaMetadata() {
        final Gson gson = new Gson();
        String jsonString = Environment.getSkjemaerJson();
        this.temaMap = gson.fromJson(jsonString, new TypeToken<HashMap<String, TemaKodeverk>>(){}.getType());
    }

    public boolean inAutoList(String tema, String skjema){
        if(temaMap.containsKey(tema)){
            return temaMap.get(tema).hasSkjema(skjema);
        }
        return false;
    }

    public int getFrist(String tema, String skjema){
        TemaKodeverk temaKodeverk = temaMap.get(tema);
        return temaKodeverk.getFristFromSkjema(skjema);
    }

    public String getOppgavetype(String tema, String skjema){
        TemaKodeverk temaKodeverk = temaMap.get(tema);
        return temaKodeverk.getOppgavetypeFromSkjema(skjema);
    }

    public boolean isIgnoreskjema(String tema, String skjema){
        return temaMap.get(tema).isIgnoreSkjema(skjema);
    }

    private class TemaKodeverk {
        private HashMap<String, SkjemaKodeverk> skjemaer;
        private List<String> ignoreSkjema;

        public boolean hasSkjema(final String skjema) {
            if(skjemaer == null || skjemaer.isEmpty()){
                return false;
            }
            return skjemaer.containsKey(Objects.requireNonNullElse(skjema, "null"));
        }

        public boolean isIgnoreSkjema(String skjema){
            if(ignoreSkjema == null || ignoreSkjema.isEmpty()){
                return false;
            }
            return ignoreSkjema.contains(skjema);
        }

        public int getFristFromSkjema(String skjema){
            return skjemaer.get(skjema).getFrist();
        }

        public String getOppgavetypeFromSkjema(String skjema){
            return skjemaer.get(skjema).getOppgavetype();
        }

        private class SkjemaKodeverk {
            private int frist;
            private String oppgavetype;

            public SkjemaKodeverk(int frist, String oppgavetype) {
                this.frist = frist;
                this.oppgavetype = oppgavetype;
            }

            public int getFrist() {
                return frist;
            }

            public String getOppgavetype() {
                return oppgavetype;
            }
        }
    }
}

File: src/main/kotlin/no/nav/helse/flex/operations/SkjemaMetadata.kt
package no.nav.helse.flex.operations

import java.util.*

object SkjemaMetadata {
    private val temaMap: HashMap<String, TemaKodeverk> = hashMapOf(
        "SYK" to TemaKodeverk(
            skjemaer = hashMapOf(
                "NAV 08-07.04D" to SkjemaKodeverk("SOK", 3),
                "NAV 08-07.04 D" to SkjemaKodeverk("SOK", 3)
            ),
            ignoreSkjema = listOf(
                "4936",
                "krav_om_fritak_fra_agp_gravid",
                "krav_om_fritak_fra_agp_kronisk",
                "soeknad_om_fritak_fra_agp_kronisk",
                "refusjonskrav_arbeidsgiverperiode_korona",
                "soeknad_om_fritak_fra_agp_gravid",
                "refusjonskrav_utestengt_arbeider_korona",
                "annuller_refusjonskrav_utestengt_arbeider_korona"
            )
        )
    )

    fun inAutoList(tema: String, skjema: String?): Boolean {
        return temaMap[tema]?.skjemaer?.containsKey(skjema) ?: false
    }

    fun getFrist(tema: String, skjema: String?): Int {
        return temaMap[tema]!!.skjemaer[skjema]!!.frist
    }

    fun getOppgavetype(tema: String, skjema: String?): String {
        return temaMap[tema]!!.skjemaer[skjema]!!.oppgavetype
    }

    fun isIgnoreskjema(tema: String, skjema: String?): Boolean {
        return temaMap[tema]?.ignoreSkjema?.contains(skjema) ?: false
    }
}

private class TemaKodeverk(
    val skjemaer: HashMap<String, SkjemaKodeverk>,
    val ignoreSkjema: List<String?>
)

private class SkjemaKodeverk(
    val oppgavetype: String,
    val frist: Int
)
